### PR TITLE
Beginnings of IDL conversion using Idol

### PR DESCRIPTION
**Pull request by cbiffle**
This converts most of the IPC interactions in the Gimlet image to use Idol.

Skipped for now:
- stm32h7-gpio: by accident
- i2c: on purpose to land this faster.

### CODE REVIEW

File: idl/user-leds.idol [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        ),
+        "led_toggle": (
+            args: {
+                "index": "usize",
+            },
+            reply: Result(
+                ok: "()",
+                err: CLike("LedError"),
+            ),
+            idempotent: true,
```
Comment (daym) [inline idl/user-leds.idol:34]: Is that one idempotent? I think not. It would toggle more or fewer times.

Comment (cbiffle) [reply idl/user-leds.idol:34]: Good catch!

Comment (cbiffle) [reply idl/user-leds.idol:34]: Fixed in latest patchset.

**Comment by cbiffle** [comment]
Amusingly the bug @daym found in the GPIO IDL is actually present in current `master` -- it's way more obvious when written in Idol.

**Review by bcantrill — APPROVED** [review]
I think this looks great!  In trying out `humility qspi`, I did hit a bug in idolatry -- and another opportunity for easy improvement:

https://github.com/oxidecomputer/idolatry/issues/1
https://github.com/oxidecomputer/idolatry/issues/2

It would be good to fix those -- but in debugging them, I got the opportunity to go through the idolatry-generated code, and I can't wait to build on this!

**Comment by bcantrill** [comment]
Some additional comments, none of which should prevent the integration:

1. It would be very helpful to be able to correlate task to interface.
2. It would be very helpful to have the `.idol` files in the build archive

It feels like both of these can be done by adding an `interface` field (say) in the `app.toml` that points to a `.idol` file, and then pass that through to the `build.rs` for tasks that have an interface.  Clearly it would be possible to be slightly less explicit/more magical, but it seems that being explicit is most in keeping with the way we have structured the `app.toml`. (If there are no objections, I may implement a prototype of this.)

**Comment by bcantrill** [comment]
> Some additional comments, none of which should prevent the integration:
> 
> 1. It would be very helpful to be able to correlate task to interface.
> 2. It would be very helpful to have the `.idol` files in the build archive
> 
> It feels like both of these can be done by adding an `interface` field (say) in the `app.toml` that points to a `.idol` file, and then pass that through to the `build.rs` for tasks that have an interface. Clearly it would be possible to be slightly less explicit/more magical, but it seems that being explicit is most in keeping with the way we have structured the `app.toml`. (If there are no objections, I may implement a prototype of this.)

Just to tie this off: I ended up implementing this in a way that didn't necessitate any changes to the `app.toml` or to the build archive.

**Comment by cbiffle** [comment]
> > Some additional comments, none of which should prevent the integration:
> > 
> > 1. It would be very helpful to be able to correlate task to interface.
> > 2. It would be very helpful to have the `.idol` files in the build archive
> > 
> > It feels like both of these can be done by adding an `interface` field (say) in the `app.toml` that points to a `.idol` file, and then pass that through to the `build.rs` for tasks that have an interface. Clearly it would be possible to be slightly less explicit/more magical, but it seems that being explicit is most in keeping with the way we have structured the `app.toml`. (If there are no objections, I may implement a prototype of this.)
> 
> Just to tie this off: I ended up implementing this in a way that didn't necessitate any changes to the `app.toml` or to the build archive.

FWIW, I actually agree with both points, and would like to do them both soon.

**Pull request merged**